### PR TITLE
Daly BMS improvements

### LINKED ISSUE #3086: BMS Daly EspHome DEV
### The problem

Ajout des tensions des cellule de 1 à 16 ne fonctionne pas, il ne remonte que la Température

### Which version of ESPHome has the issue?

V 2022.2.5-dev

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2022.2.9

### What platform are you using?

ESP8266

### Board

Wemos D1 mini

### Component causing the issue

Daly BMS

### Example YAML snippet

```yaml
esphome:
  name: bmsdaly

esp8266:
  board: esp01_1m

# Enable logging
logger:

# Enable Home Assistant API
api:

ota:
  password: "057e5e176a11c4f5ac1cab965ff9bf6c"

wifi:
  ssid: !secret ssid
  password: !secret mp_wifi


  # Enable fallback hotspot (captive portal) in case wifi connection fails
  ap:
    ssid: "Bmsdaly Fallback Hotspot"
    password: "MzJE8fEg3jGy"

captive_portal:


uart:
  tx_pin: GPIO15
  rx_pin: GPIO13
  baud_rate: 9600
  
switch:
  - platform: restart
    name: "bms_daly_restart"
    
daly_bms:
  update_interval: 20s

sensor:
  - platform: daly_bms
    voltage:
      name: "Battery Voltage"
    current:
      name: "Battery Current"
    battery_level:
      name: "Battery Level"
    max_cell_voltage:
      name: "Max Cell Voltage"
    max_cell_voltage_number:
      name: "Max Cell Voltage Number"
    min_cell_voltage:
      name: "Min Cell Voltage"
    min_cell_voltage_number:
      name: "Min Cell Voltage Number"
    max_temperature:
      name: "Max Temperature"
    max_temperature_probe_number:
      name: "Max Temperature Probe Number"
    min_temperature:
      name: "Min Temperature"
    min_temperature_probe_number:
      name: "Min Temperature Probe Number"
    remaining_capacity:
      name: "Remaining Capacity"
      filters:
      - lambda: return x * 0.04;
    cells_number:
      name: "Cells Number"
    temperature_1:
      name: "Temperature 1"
    temperature_2:
      name: "Temperature 2"
    
 # - plateform: template
      #name: "Capacité en KW"
      #remaining_capacity*0.04
      
      
      
#intensité du signal wifi
  - platform: wifi_signal
    name: "WiFi puissance_BMS_Daly"
    update_interval: 60s

text_sensor:
  - platform: daly_bms
    status:
      name: "BMS Status"


binary_sensor:
  - platform: daly_bms
    charging_mos_enabled:
      name: "Charging MOS"
    discharging_mos_enabled:
      name: "Discharging MOS"
```

### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/esphome/bmsdaly.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing bmsdaly (board: esp01_1m; framework: arduino; platform: platformio/espressif8266 @ 3.2.0)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 1MB Flash
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- <ESPAsyncTCP-esphome> 1.2.3
|-- <ESPAsyncWebServer-esphome> 2.1.0
|   |-- <ESPAsyncTCP-esphome> 1.2.3
|   |-- <Hash> 1.0
|   |-- <ESP8266WiFi> 1.0
|-- <DNSServer> 1.1.1
|-- <ESP8266WiFi> 1.0
|-- <ESP8266mDNS> 1.2
Compiling /data/bmsdaly/.pioenvs/bmsdaly/src/esphome/components/debug/debug_component.cpp.o
Compiling /data/bmsdaly/.pioenvs/bmsdaly/src/main.cpp.o
Linking /data/bmsdaly/.pioenvs/bmsdaly/firmware.elf
RAM:   [====      ]  41.6% (used 34092 bytes from 81920 bytes)
Flash: [====      ]  42.0% (used 430573 bytes from 1023984 bytes)
Building /data/bmsdaly/.pioenvs/bmsdaly/firmware.bin
esp8266_copy_factory_bin(["/data/bmsdaly/.pioenvs/bmsdaly/firmware.bin"], ["/data/bmsdaly/.pioenvs/bmsdaly/firmware.elf"])
========================= [SUCCESS] Took 20.19 seconds =========================
INFO Successfully compiled program.
INFO Resolving IP address of bmsdaly.local
INFO  -> 192.168.1.28
INFO Uploading /data/bmsdaly/.pioenvs/bmsdaly/firmware.bin (434720 bytes)
INFO Compressed to 299534 bytes
Uploading: [============================================================] 100% Done...

INFO Waiting for result...
INFO OTA successful
INFO Successfully uploaded program.
INFO Starting log output from bmsdaly.local using esphome API
INFO Successfully connected to bmsdaly.local
[11:26:44][I][app:102]: ESPHome version 2022.3.0-dev compiled on Feb 23 2022, 11:26:03
[11:26:44][C][wifi:491]: WiFi:
[11:26:44][C][wifi:353]:   Local MAC: 40:91:51:4F:8F:9E
[11:26:44][C][wifi:354]:   SSID: 'Livebox-596E'[redacted]
[11:26:44][C][wifi:355]:   IP Address: 192.168.1.28
[11:26:44][C][wifi:356]:   BSSID: E8:1B:69:02:59:6E[redacted]
[11:26:44][C][wifi:358]:   Hostname: 'bmsdaly'
[11:26:44][C][wifi:360]:   Signal strength: -51 dB ▂▄▆█
[11:26:44][C][wifi:364]:   Channel: 1
[11:26:44][C][wifi:365]:   Subnet: 255.255.255.0
[11:26:44][C][wifi:366]:   Gateway: 192.168.1.1
[11:26:44][C][wifi:367]:   DNS1: 192.168.1.1
[11:26:44][C][wifi:368]:   DNS2: 0.0.0.0
[11:26:44][C][logger:233]: Logger:
[11:26:44][C][logger:234]:   Level: DEBUG
[11:26:44][C][logger:235]:   Log Baud Rate: 115200
[11:26:44][C][logger:236]:   Hardware UART: UART0
[11:26:44][C][uart.arduino_esp8266:102]: UART Bus:
[11:26:44][C][uart.arduino_esp8266:103]:   TX Pin: GPIO15
[11:26:44][C][uart.arduino_esp8266:104]:   RX Pin: GPIO13
[11:26:44][C][uart.arduino_esp8266:106]:   RX Buffer Size: 256
[11:26:44][C][uart.arduino_esp8266:108]:   Baud Rate: 9600 baud
[11:26:44][C][uart.arduino_esp8266:109]:   Data Bits: 8
[11:26:44][C][uart.arduino_esp8266:110]:   Parity: NONE
[11:26:44][C][uart.arduino_esp8266:111]:   Stop bits: 1
[11:26:44][C][uart.arduino_esp8266:115]:   Using software serial
[11:26:44][C][restart:022]: Restart Switch 'bms_daly_restart'
[11:26:44][C][restart:022]:   Icon: 'mdi:restart'
[11:26:44][C][daly_bms:025]: Daly BMS:
[11:26:44][C][captive_portal:088]: Captive Portal:
[11:26:44][C][mdns:084]: mDNS:
[11:26:44][C][mdns:085]:   Hostname: bmsdaly
[11:26:44][C][ota:085]: Over-The-Air Updates:
[11:26:44][C][ota:086]:   Address: bmsdaly.local:8266
[11:26:44][C][ota:089]:   Using Password.
[11:26:44][C][api:138]: API Server:
[11:26:44][C][api:139]:   Address: bmsdaly.local:6053
[11:26:44][C][api:143]:   Using noise encryption: NO
[11:26:44][C][wifi_signal.sensor:009]: WiFi Signal 'WiFi puissance_BMS_Daly'
[11:26:44][C][wifi_signal.sensor:009]:   Device Class: 'signal_strength'
[11:26:44][C][wifi_signal.sensor:009]:   State Class: 'measurement'
[11:26:44][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dBm'
[11:26:44][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0
[11:26:44][C][debug:049]: Debug component:
[11:26:44][C][debug:051]:   Device info 'Device Info'
[11:26:44][C][debug:054]:   Free space on heap 'Heap Free'
[11:26:44][C][debug:054]:     State Class: ''
[11:26:44][C][debug:054]:     Unit of Measurement: 'B'
[11:26:44][C][debug:054]:     Accuracy Decimals: 0
[11:26:44][C][debug:054]:     Icon: 'mdi:counter'
[11:26:44][C][debug:055]:   Largest free heap block 'Heap Max Block'
[11:26:44][C][debug:055]:     State Class: ''
[11:26:44][C][debug:055]:     Unit of Measurement: 'B'
[11:26:44][C][debug:055]:     Accuracy Decimals: 0
[11:26:44][C][debug:055]:     Icon: 'mdi:counter'
[11:26:44][C][debug:057]:   Heap fragmentation 'Heap Fragmentation'
[11:26:44][C][debug:057]:     State Class: ''
[11:26:44][C][debug:057]:     Unit of Measurement: '%'
[11:26:44][C][debug:057]:     Accuracy Decimals: 1
[11:26:44][C][debug:057]:     Icon: 'mdi:counter'
[11:26:44][D][debug:061]: ESPHome version 2022.3.0-dev
[11:26:44][D][debug:065]: Free Heap Size: 25440 bytes
[11:26:44][D][debug:093]: Flash Chip: Size=1024kB Speed=40MHz Mode=DOUT
[11:26:44][D][debug:250]: Chip ID: 0x004F8F9E
[11:26:44][D][debug:251]: SDK Version: 2.2.2-dev(38a443e)
[11:26:44][D][debug:252]: Core Version: 3.0.2
[11:26:44][D][debug:253]: Boot Version=31 Mode=1
[11:26:44][D][debug:254]: CPU Frequency: 80
[11:26:44][D][debug:255]: Flash Chip ID=0x00164020
[11:26:44][D][debug:256]: Reset Reason: Software/System restart
[11:26:44][D][debug:257]: Reset Info: Software/System restart
[11:26:45][D][text_sensor:067]: 'Device Info': Sending state '2022.3.0-dev|Flash: 1024kB Speed:40MHz Mode:DOUT|Chip: 0x004f8f9e|SDK: 2.2.2-dev(38a443e)|Core: 3.0.2|Boot: 31|Mode: 1|CPU: 80|Flash: 0x00164020|Reset: Software/System restart|Software/System restart'
[11:26:48][D][sensor:124]: 'Heap Free': Sending state 27992.00000 B with 0 decimals of accuracy
[11:26:48][D][sensor:124]: 'Heap Max Block': Sending state 23576.00000 B with 0 decimals of accuracy
[11:26:48][D][sensor:124]: 'Heap Fragmentation': Sending state 6.00000 % with 1 decimals of accuracy
[11:26:48][D][sensor:124]: 'Loop Time': Sending state 165.00000 ms with 0 decimals of accuracy
[11:26:48][D][sensor:124]: 'Temperature 1': Sending state 13.00000 °C with 0 decimals of accuracy
[11:26:48][D][sensor:124]: 'Temperature 2': Sending state -40.00000 °C with 0 decimals of accuracy
[11:26:53][D][sensor:124]: 'Heap Free': Sending state 28832.00000 B with 0 decimals of accuracy
[11:26:53][D][sensor:124]: 'Heap Max Block': Sending state 24376.00000 B with 0 decimals of accuracy
[11:26:53][D][sensor:124]: 'Heap Fragmentation': Sending state 6.00000 % with 1 decimals of accuracy
[11:26:53][D][sensor:124]: 'Loop Time': Sending state 124.00000 ms with 0 decimals of accuracy
[11:26:58][D][sensor:124]: 'Heap Free': Sending state 28520.00000 B with 0 decimals of accuracy
[11:26:58][D][sensor:124]: 'Heap Max Block': Sending state 24376.00000 B with 0 decimals of accuracy
[11:26:58][D][sensor:124]: 'Heap Fragmentation': Sending state 5.00000 % with 1 decimals of accuracy
[11:26:58][D][sensor:124]: 'Loop Time': Sending state 36.00000 ms with 0 decimals of accuracy
[11:27:03][D][sensor:124]: 'Heap Free': Sending state 28952.00000 B with 0 decimals of accuracy
[11:27:03][D][sensor:124]: 'Heap Max Block': Sending state 24376.00000 B with 0 decimals of accuracy
[11:27:03][D][sensor:124]: 'Heap Fragmentation': Sending state 6.00000 % with 1 decimals of accuracy
[11:27:03][D][sensor:124]: 'Loop Time': Sending state 35.00000 ms with 0 decimals of accuracy
```

### Additional information

Fonctionne avec la version Esphome avec Daly BMS:

NFO Reading configuration /config/esphome/bmsdaly.yaml...
INFO Starting log output from bmsdaly.local using esphome API
INFO Successfully connected to bmsdaly.local
[21:01:36][I][app:102]: ESPHome version 2022.2.4 compiled on Feb 23 2022, 11:29:20
[21:01:36][C][wifi:491]: WiFi:
[21:01:36][C][wifi:353]:   Local MAC: 40:91:51:4F:8F:9E
[21:01:36][C][wifi:354]:   SSID: 'Livebox-596E'[redacted]
[21:01:36][C][wifi:355]:   IP Address: 192.168.1.28
[21:01:36][C][wifi:356]:   BSSID: E8:1B:69:02:59:6E[redacted]
[21:01:36][C][wifi:358]:   Hostname: 'bmsdaly'
[21:01:36][C][wifi:360]:   Signal strength: -54 dB ▂▄▆█
[21:01:36][C][wifi:364]:   Channel: 1
[21:01:36][C][wifi:365]:   Subnet: 255.255.255.0
[21:01:36][C][wifi:366]:   Gateway: 192.168.1.1
[21:01:36][C][wifi:367]:   DNS1: 192.168.1.1
[21:01:36][C][wifi:368]:   DNS2: 0.0.0.0
[21:01:36][C][logger:233]: Logger:
[21:01:36][C][logger:234]:   Level: DEBUG
[21:01:36][C][logger:235]:   Log Baud Rate: 115200
[21:01:36][C][logger:236]:   Hardware UART: UART0
[21:01:36][C][uart.arduino_esp8266:102]: UART Bus:
[21:01:36][C][uart.arduino_esp8266:103]:   TX Pin: GPIO15
[21:01:36][C][uart.arduino_esp8266:104]:   RX Pin: GPIO13
[21:01:36][C][uart.arduino_esp8266:106]:   RX Buffer Size: 256
[21:01:36][C][uart.arduino_esp8266:108]:   Baud Rate: 9600 baud
[21:01:36][C][uart.arduino_esp8266:109]:   Data Bits: 8
[21:01:36][C][uart.arduino_esp8266:110]:   Parity: NONE
[21:01:36][C][uart.arduino_esp8266:111]:   Stop bits: 1
[21:01:36][C][uart.arduino_esp8266:115]:   Using software serial
[21:01:36][C][restart:022]: Restart Switch 'bms_daly_restart'
[21:01:36][C][restart:022]:   Icon: 'mdi:restart'
[21:01:36][C][daly_bms:024]: Daly BMS:
[21:01:36][C][captive_portal:144]: Captive Portal:
[21:01:36][C][mdns:084]: mDNS:
[21:01:36][C][mdns:085]:   Hostname: bmsdaly
[21:01:36][C][ota:085]: Over-The-Air Updates:
[21:01:36][C][ota:086]:   Address: bmsdaly.local:8266
[21:01:36][C][ota:089]:   Using Password.
[21:01:36][C][api:138]: API Server:
[21:01:36][C][api:139]:   Address: bmsdaly.local:6053
[21:01:36][C][api:143]:   Using noise encryption: NO
[21:01:36][C][wifi_signal.sensor:009]: WiFi Signal 'WiFi puissance_BMS_Daly'
[21:01:36][C][wifi_signal.sensor:009]:   Device Class: 'signal_strength'
[21:01:36][C][wifi_signal.sensor:009]:   State Class: 'measurement'
[21:01:36][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dBm'
[21:01:36][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0
[21:01:43][D][sensor:124]: 'Max Cell Voltage': Sending state 3.32200 V with 2 decimals of accuracy
[21:01:43][D][sensor:124]: 'Max Cell Voltage Number': Sending state 13.00000  with 0 decimals of accuracy
[21:01:43][D][sensor:124]: 'Min Cell Voltage': Sending state 3.31500 V with 2 decimals of accuracy
[21:01:43][D][sensor:124]: 'Min Cell Voltage Number': Sending state 1.00000  with 0 decimals of accuracy
[21:01:43][D][sensor:124]: 'Max Temperature': Sending state 13.00000 °C with 0 decimals of accuracy
[21:01:43][D][sensor:124]: 'Max Temperature Probe Number': Sending state 1.00000  with 0 decimals of accuracy
[21:01:43][D][sensor:124]: 'Min Temperature': Sending state 13.00000 °C with 0 decimals of accuracy
[21:01:43][D][sensor:124]: 'Min Temperature Probe Number': Sending state 1.00000  with 0 decimals of accuracy
[21:01:43][D][text_sensor:067]: 'BMS Status': Sending state 'Discharging'
[21:01:43][D][sensor:124]: 'Remaining Capacity': Sending state 6.93360 Ah with 2 decimals of accuracy
[21:01:43][D][sensor:124]: 'Cells Number': Sending state 16.00000  with 0 decimals of accuracy
[21:01:43][D][sensor:124]: 'Temperature 1': Sending state 13.00000 °C with 0 decimals of accuracy
[21:01:43][D][sensor:124]: 'Temperature 2': Sending state -40.00000 °C with 0 decimals of accuracy
[21:01:47][D][sensor:124]: 'WiFi puissance_BMS_Daly': Sending state -56.00000 dBm with 0 decimals of accuracy

### PR DESCRIPTION
**Pull request by matthias882**
# What does this implement/fix?

Issue https://github.com/esphome/issues/issues/3086: Not all Data fom BMS could be received:
Problem was stuck in RX because of sending all requests to DalyBMS at one Time.
Solution: Buildet a state-machine in loop to sende one request after the other. First request triggert by update.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/3086

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/daly_bms/daly_bms.h [hunk 1]
```diff
@@ -110,6 +111,13 @@ class DalyBmsComponent : public PollingComponent, public uart::UARTDevice {
 
   binary_sensor::BinarySensor *charging_mos_enabled_{nullptr};
   binary_sensor::BinarySensor *discharging_mos_enabled_{nullptr};
+
+  std::vector<uint8_t> data_;
+  bool receiving_{false};
+  uint8_t data_count_;
+  uint32_t last_transmission_{0};
+  bool trigger_next;
+  uint8_t next_request;
```
Comment (jesserockz) [inline esphome/components/daly_bms/daly_bms.h]: trailing underscore for protected members, you will need to update them manually in the cpp file too
```suggestion
  bool trigger_next_;
  uint8_t next_request_;
```

File: esphome/components/daly_bms/daly_bms.cpp [hunk 2]
```diff
@@ -19,29 +19,90 @@ static const uint8_t DALY_REQUEST_STATUS = 0x94;
...
+                next_request = 6;
+            break;
+            case 6:
+                this->request_data_(DALY_REQUEST_TEMPERATURE);
+                next_request = 7;
+            break;
+            case 7:
+            break;
+            default:
+            break;
```
Comment (jesserockz) [inline esphome/components/daly_bms/daly_bms.cpp]:
```suggestion
            case 7:
            default:
            break;
```

File: esphome/components/daly_bms/daly_bms.cpp [hunk 3]
```diff
@@ -19,29 +19,90 @@ static const uint8_t DALY_REQUEST_STATUS = 0x94;
...
+      // last transmission too long ago. Reset RX index.
+      ESP_LOGW(TAG, "Last transmission too long ago. Reset RX index.");
+      data_.clear();
+      receiving_ = false;
+    }
+    if ((now - last_transmission_ >= 250) && (trigger_next == false)) {
+        //last transmittion loger than 0.25s ago -> trigger next request
+        last_transmission_ = now;
+        trigger_next = true;
+    }
```
Comment (jesserockz) [inline esphome/components/daly_bms/daly_bms.cpp]:
```suggestion
    if ((now - last_transmission_ >= 250) && !trigger_next) {
      //last transmittion loger than 0.25s ago -> trigger next request
      last_transmission_ = now;
      trigger_next = true;
    }
```

File: esphome/components/daly_bms/daly_bms.cpp [hunk 4]
```diff
@@ -19,29 +19,90 @@ static const uint8_t DALY_REQUEST_STATUS = 0x94;
...
+        }
+        data_.push_back(c);
+        if (data_.size() == 4)
+            data_count_ = c;
+        if ((data_.size() > 4) and (data_.size() == data_count_ + 5)) {
+          this->decode_data_(data_);
+          data_.clear();
+          receiving_ = false;
+        }
+    }
```
Comment (jesserockz) [inline esphome/components/daly_bms/daly_bms.cpp]:
```suggestion
    while (available()) {
      uint8_t c;
      read_byte(&c);
      if (!receiving_) {
        if (c != 0xa5)
          continue;
        receiving_ = true;
      }
      data_.push_back(c);
      if (data_.size() == 4)
          data_count_ = c;
      if ((data_.size() > 4) and (data_.size() == data_count_ + 5)) {
        this->decode_data_(data_);
        data_.clear();
        receiving_ = false;
      }
    }
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
My review only addresses the formatting issues for now

**Comment by matthias882** [comment]
The formatting should be pretty now...

**Comment by h3po** [comment]
The code works on my esp32 as well. But with or without the patch, I'm getting this warning message:
```
[21:19:16][V][component:199]: Component daly_bms took a long time for an operation (0.11 s).
[21:19:16][V][component:200]: Components should block for at most 20-30ms.
```

**Comment by frenck** [comment]
@matthias882 There is still an CI error happening. Could you take a look?

**Comment by h3po** [comment]
> @matthias882 There is still an CI error happening. Could you take a look?
I don't see a ci error, it's just @jesserockz is marked as reviewer for the formatting changes

File: esphome/components/daly_bms/daly_bms.cpp [hunk 5]
```diff
@@ -237,6 +298,9 @@ void DalyBmsComponent::decode_data_(std::vector<uint8_t> data) {
             break;
         }
       }
+      else {
+          ESP_LOGW(TAG,"Checksum-Error on Packet %x", it[4]);
+      }
```
Comment (jesserockz) [inline esphome/components/daly_bms/daly_bms.cpp]:
```suggestion
      } else {
        ESP_LOGW(TAG,"Checksum-Error on Packet %x", it[4]);
      }
```

File: esphome/components/daly_bms/daly_bms.cpp [hunk 6]
```diff
@@ -19,28 +19,88 @@ static const uint8_t DALY_REQUEST_STATUS = 0x94;
...
+    data_.clear();
+    receiving_ = false;
+  }
+  if ((now - last_transmission_ >= 250) && !trigger_next_) {
+    //last transmittion loger than 0.25s ago -> trigger next request
+    last_transmission_ = now;
+    trigger_next_ = true;
+  }
+  if (available())
+  last_transmission_ = now;
```
Comment (jesserockz) [inline esphome/components/daly_bms/daly_bms.cpp]:
```suggestion
    last_transmission_ = now;
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by matthias882** [comment]
Thanks for nothing

**Comment by jesserockz** [comment]
Hi @matthias882 Apologies for the very long delay on this PR. 

I have pushed some extra changes into the branch that should not change the functionality, but adjusts the code style slightly.

**Comment by matthias882** [comment]
Hi @jesserockz 
I tried to put the new code to my daly-bms-controller to test it, but the compilation stopped with several errors...

```
INFO Reading configuration /opt/iobroker/iobroker-data/esphome.0/daly-smartbms-1.yaml...
INFO Updating https://github.com/matthias882/esphome.git@patch-4
INFO Generating C++ source...
INFO Compiling app...
Processing daly-smartbms-1 (board: d1_mini; framework: arduino; platform: platformio/espressif8266 @ 3.2.0)
--------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
Dependency Graph
|-- ESPAsyncTCP-esphome @ 1.2.3
|-- ESPAsyncWebServer-esphome @ 2.1.0
|   |-- ESPAsyncTCP-esphome @ 1.2.3
|   |-- Hash @ 1.0
|   |-- ESP8266WiFi @ 1.0
|-- DNSServer @ 1.1.1
|-- ESP8266WiFi @ 1.0
|-- ESP8266mDNS @ 1.2
|-- ArduinoJson @ 6.18.5
Compiling .pioenvs/daly-smartbms-1/src/esphome/components/daly_bms/daly_bms.cpp.o
Compiling .pioenvs/daly-smartbms-1/src/main.cpp.o
In file included from src/esphome/components/daly_bms/daly_bms.cpp:1:
src/esphome/components/daly_bms/daly_bms.h:26:14: error: 'voltage' has not been declared
   26 |   SUB_SENSOR(voltage);
      |              ^~~~~~~
src/esphome/components/daly_bms/daly_bms.h:26:3: error: ISO C++ forbids declaration of 'SUB_SENSOR' with no type [-fpermissive]
   26 |   SUB_SENSOR(voltage);
      |   ^~~~~~~~~~
src/esphome/components/daly_bms/daly_bms.h:27:14: error: 'current' has not been declared
   27 |   SUB_SENSOR(current);
      |              ^~~~~~~
src/esphome/components/daly_bms/daly_bms.h:27:3: error: ISO C++ forbids declaration of 'SUB_SENSOR' with no type [-fpermissive]
   27 |   SUB_SENSOR(current);
      |   ^~~~~~~~~~
src/esphome/components/daly_bms/daly_bms.h:27:3: error: 'int esphome::daly_bms::DalyBmsComponent::SUB_SENSOR(int)' cannot be overloaded with 'int esphome::daly_bms::DalyBmsComponent::SUB_SENSOR(int)'
src/esphome/components/daly_bms/daly_bms.h:26:3: note: previous declaration 'int esphome::daly_bms::DalyBmsComponent::SUB_SENSOR(int)'
   26 |   SUB_SENSOR(voltage);
      |   ^~~~~~~~~~
src/esphome/components/daly_bms/daly_bms.h:28:14: error: 'battery_level' has not been declared
   28 |   SUB_SENSOR(battery_level);
      |              ^~~~~~~~~~~~~
```

... and so on

**Comment by jesserockz** [comment]
Oops, fixed by removing the semi colons that were auto added =)

**Review by jesserockz — APPROVED** [review]

**Comment by matthias882** [comment]
Oh no, bad news. The semi colon was not the problem. The errors are the same now, but in addition that a semi colon would be expected where you removed it ;)

```
Compiling .pioenvs/daly-smartbms-1/src/esphome/components/daly_bms/daly_bms.cpp.o
Compiling .pioenvs/daly-smartbms-1/src/main.cpp.o
In file included from src/esphome/components/daly_bms/daly_bms.cpp:1:
src/esphome/components/daly_bms/daly_bms.h:26:14: error: 'voltage' has not been declared
   26 |   SUB_SENSOR(voltage)
      |              ^~~~~~~
src/esphome/components/daly_bms/daly_bms.h:26:3: error: ISO C++ forbids declaration of 'SUB_SENSOR' with no type [-fpermissive]
   26 |   SUB_SENSOR(voltage)
      |   ^~~~~~~~~~
src/esphome/components/daly_bms/daly_bms.h:26:21: error: expected ';' at end of member declaration
   26 |   SUB_SENSOR(voltage)
      |                     ^
      |                      ;
src/esphome/components/daly_bms/daly_bms.h:27:14: error: 'current' has not been declared
   27 |   SUB_SENSOR(current)
      |              ^~~~~~~
src/esphome/components/daly_bms/daly_bms.h:27:3: error: ISO C++ forbids declaration of 'SUB_SENSOR' with no type [-fpermissive]
   27 |   SUB_SENSOR(current)
      |   ^~~~~~~~~~
src/esphome/components/daly_bms/daly_bms.h:27:21: error: expected ';' at end of member declaration
   27 |   SUB_SENSOR(current)
      |                     ^
      |                      ;
src/esphome/components/daly_bms/daly_bms.h:27:3: error: 'int esphome::daly_bms::DalyBmsComponent::SUB_SENSOR(int)' cannot be overloaded with 'int esphome::daly_bms::DalyBmsComponent::SUB_SENSOR(int)'
   27 |   SUB_SENSOR(current)
      |   ^~~~~~~~~~
src/esphome/components/daly_bms/daly_bms.h:26:3: note: previous declaration 'int esphome::daly_bms::DalyBmsComponent::SUB_SENSOR(int)'
   26 |   SUB_SENSOR(voltage)
      |   ^~~~~~~~~~```

**Comment by matthias882** [comment]
The Problem should be on my side. I think the "SUB_SENSOR()" macro in the sensor.h is perhaps to new and not in my (beta state) ESPHOME-Integration in iobroker. I try to update this and then we will see.
PS: great idea with this marco, it saves much lines and much copy/paste on creating new sensor-values!

**Comment by jesserockz** [comment]
`SUB_SENSOR` was added many months ago now and released in 2023.2.0, if not using HA you should just run ESPHome locally on computer, or as a docker container to get the latest updates.

**Comment by matthias882** [comment]
Yes, that's the problem. My esphome is currently on v2022.10.0😅 I currently try to update and then let's see...

**Pull request merged**